File: python3/middle_of_the_linked_list.py
from .model import ListNode

"""
Space   : O(1)
Time    : O(n)
"""
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow, fast = head, head

        while fast.next:
            slow = slow.next
            fast = fast.next

            if fast and fast.next:
                fast = fast.next

        return slow

File: golang/middle_of_the_linked_list.go
package main

// Space   : O(n)
// Time    : O(n)

import "golang/model"

func middleNode(head *model.ListNode) *model.ListNode {
	p1, p2 := head, head
	l := 0

	for p1 != nil {
		l++
		p1 = p1.Next
	}

	l /= 2
	for l > 0 {
		p2 = p2.Next
		l--
	}

	return p2
}
